src/Analyzers/CSharp/Analyzers/ConvertProgram/ConvertToProgramMainDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+#if CODE_STYLE
+using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
+#endif
+
+namespace Microsoft.CodeAnalysis.CSharp.TopLevelStatements
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class ConvertToProgramMainDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public ConvertToProgramMainDiagnosticAnalyzer()
+            : base(
+                  IDEDiagnosticIds.UseProgramMainId,
+                  EnforceOnBuildValues.UseProgramMain,
+                  CSharpCodeStyleOptions.PreferTopLevelStatements,
+                  LanguageNames.CSharp,
+                  new LocalizableResourceString(nameof(CSharpAnalyzersResources.Convert_to_Program_Main_style_program), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(ProcessCompilationUnit, SyntaxKind.CompilationUnit);
+
+        private void ProcessCompilationUnit(SyntaxNodeAnalysisContext context)
+        {
+            var options = context.Options;
+            var root = (CompilationUnitSyntax)context.Node;
+            var syntaxTree = root.SyntaxTree;
+
+            var cancellationToken = context.CancellationToken;
+
+            if (!HasGlobalStatement(root))
+                return;
+
+            var optionSet = options.GetAnalyzerOptionSet(syntaxTree, cancellationToken);
+            var option = optionSet.GetOption(CSharpCodeStyleOptions.PreferTopLevelStatements);
+
+            // if they prefer top level statements, there's nothing for us to do as this code matches their preference
+            if (option.Value)
+                return;
+
+            var compilation = context.Compilation;
+
+            // resiliency check for later on.  This shouldn't happen but we don't want to crash if we are in a weird
+            // state where we have top level statements but no 'Program' type.
+            var programType = compilation.GetBestTypeByMetadataName(WellKnownMemberNames.TopLevelStatementsEntryPointTypeName);
+            if (programType == null)
+                return;
```

Comment:
Note: This actually can happen if the user is using new analyzers package with old compiler that used to have unspeakable name `<Program>$` instead of `Program` (https://github.com/dotnet/roslyn/pull/55368)

There is nothing to do for that. Just wanted to note :)

Reply:
yeah, don't care :)  that wasn't an official shipped version of the language.